drivers/mfd/mfd_ite_it8801.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2024 ITE Corporation. All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ite_it8801_mfd
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/mfd/mfd_ite_it8801.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mfd_ite_it8801, CONFIG_MFD_LOG_LEVEL);
+
+static const struct device *kbd_dev = DEVICE_DT_GET_ANY(ite_it8801_kbd);
+
+struct mfd_it8801_config {
+	const struct i2c_dt_spec i2c_dev;
+	/* Alert GPIO pin */
+	const struct gpio_dt_spec irq_gpios;
+	/* MFD sub device */
+	const struct device **sub_mfd_dev;
+	/* MFD sub device count */
+	uint8_t sub_mfd_node_num;
+};
+
+struct mfd_it8801_data {
+	struct k_work gpio_isr_worker;
+	/* Alert pin callback */
+	struct gpio_callback gpio_cb;
+	const struct device *mfd_dev;
+};
+
+const struct i2c_dt_spec *mfd_it8801_get_i2c_dt_spec(const struct device *dev)
+{
+	const struct mfd_it8801_config *config = dev->config;
+
+	return &config->i2c_dev;
+}
+
+static int it8801_check_vendor_id(const struct device *dev)
+{
+	const struct mfd_it8801_config *config = dev->config;
+	int i, ret;
+	uint8_t val;
+
+	/*  Verify vendor ID registers(16-bits). */
+	for (i = 0; i < ARRAY_SIZE(it8801_id_verify); i++) {
+		ret = i2c_reg_read_byte_dt(&config->i2c_dev, it8801_id_verify[i].reg, &val);
+
+		if (ret != 0) {
+			LOG_ERR("Failed to read vendoer ID (ret %d)", ret);
+			return ret;
+		}
+
+		if (val != it8801_id_verify[i].chip_id) {
+			LOG_ERR("The IT8801 vendor ID is wrong");
+			return -ENODEV;
+		}
+	}
+
+	return ret;
+}
+
+void it8801_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	ARG_UNUSED(pins);
+	struct mfd_it8801_data *data = CONTAINER_OF(cb, struct mfd_it8801_data, gpio_cb);
+
+	k_work_submit(&data->gpio_isr_worker);
+}
+
+static void it8801_gpio_alert_worker(struct k_work *work)
+{
+	struct mfd_it8801_data *data = CONTAINER_OF(work, struct mfd_it8801_data, gpio_isr_worker);
+	const struct mfd_it8801_config *config = data->mfd_dev->config;
+	int i;
+
+	for (i = 0; i < config->sub_mfd_node_num; i++) {
+		if (config->sub_mfd_dev[i] != NULL) {
+			if (config->sub_mfd_dev[i] == kbd_dev) {
+				it8801_input_alert_handler(config->sub_mfd_dev[i]);
```

Comment:
This function isn't available in this commit.  Each commit should build on it's own.   Add this support with the KBD driver commit.

You could have the child drivers register an alert callback.  Then MFD driver doesn't need to know the type of each child.

If you wanted the alert callbacks to be configured at build time, you could define an iterable section per it8801 instance.

See `INPUT_CALL_DEFINE` for an example of creating an iterable section for a specific device instance.

Reply:
Moved to respective driver.

Because the KBD and GPIO devices of IT8801 share the gpio irq. Consider that if IT8801 is connected to the EC, the GPIO interrupt is enabled in the interrupt configuration using irq_connect_dynamic.
Repeated registration of gpio irq will cause [Assert ](https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/common/dynamic_isr.c#L23)problem, so here we only register it once in mfd_ite_it8801.